hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/KafkaSourceUtil.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.utilities.sources.helpers;
+
+import org.apache.hudi.common.config.TypedProperties;
+import org.apache.hudi.common.util.StringUtils;
+import org.apache.hudi.common.util.hash.HashID;
+import org.apache.hudi.utilities.exception.HoodieReadFromSourceException;
+import org.apache.hudi.utilities.schema.SchemaProvider;
+
+import com.google.crypto.tink.subtle.Base64;
+
+import java.util.Objects;
+
+import static org.apache.hudi.utilities.config.KafkaSourceConfig.KAFKA_VALUE_DESERIALIZER_SCHEMA;
+
+public class KafkaSourceUtil {
+
+  public static final String NATIVE_KAFKA_CONSUMER_GROUP_ID = "group.id";
+  public static final int GROUP_ID_MAX_BYTES_LENGTH = 255;
+
+  public static void configureSchemaDeserializer(SchemaProvider schemaProvider, TypedProperties props) {
+    if (schemaProvider == null || Objects.isNull(schemaProvider.getSourceSchema())) {
+      throw new HoodieReadFromSourceException("SchemaProvider has to be set to use KafkaAvroSchemaDeserializer");
+    }
+    props.put(KAFKA_VALUE_DESERIALIZER_SCHEMA.key(), schemaProvider.getSourceSchema().toString());
+    // assign consumer group id based on the schema, since if there's a change in the schema we ensure KafkaRDDIterator doesn't use cached Kafka Consumer
+    String groupId = props.getString(NATIVE_KAFKA_CONSUMER_GROUP_ID, "");
```

Comment:
Instead of setting up the group id, why not just fetch the latest schema and set up it again if the schema evvolved.

Reply:
The KafkaRDD class from the Kafka client package is caching the KafkaConsumer for a given group id and partition. As a result, we are using KafkaAvroSchemaDeserializer with the previous schema to deserialize the records. This leads to issues during schema evolution.

Reply:
Does Kafka client itself has an option to control the cache TTL?

Reply:
No, while creating the Kafka RDD, it's always hardcoded to true
https://github.com/apache/spark/blob/876450c4130062a80f4ecb7b7afd232b5481a99f/connector/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/KafkaUtils.scala#L70

Since useConsumerCache is set to true. We always get the existing consumer with outdated schema

Reply:
Can we copy this utility into hudi and override the flag as false then?

Reply:
Yeah, I think that should be possible.

Reply:
But if we override the flat to false, we might end up creating new KafkaConsumer every time.

Reply:
Can we create based on the schema change then? The delta-streamer is a micro-batch for each run, maybe we can add some smart strategy here.

Reply:
This change will do exactly the same. Caching the KafkaConsumer happens as part of the RDD creation, I'm not sure we can extract that functionality of the kafka utils libr.

Reply:
> Caching the KafkaConsumer happens as part of the RDD creation

Does it make sense if we hard code the cache as false first then create/reuse the consumer based on the schema change?

Reply:
We can't control how the KafkaConsumer is created, It's done during the RDD creation. I don't think we can control that.